[24.11]: workflows/labels: skip for staging-next / haskell-updates / python-updates

wolfgangwalther opened the pull request:
Manual backport of #402332 and parts of #374921.

I did not backport the configuration of backport labels for CI PRs from #374921... because otherwise we'd end up with backport labels on the backport PRs as well. On the same vein, we'll need to remove those labels from release-25.05 after branch-off.

## Things done

- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

pbsds committed:
workflows/labels: add a `sync-labels: false` step, migrate some rules

(cherry picked from commit 8ed4f7b5a62b2e41606eb66b4c5f6d142f46370b)

wolfgangwalther committed:
workflows/labels: skip for staging-next / haskell-updates / python-updates

Labelling those PRs is more annoying then useful.

(cherry picked from commit 19095200155c0164eeb1629d3ad5f2a53ad40e15)

wolfgangwalther committed:
labels: add workflow related labels

This adds some labels related to the higher level workflows of
developing nixpkgs:
- "4.workflow: backport" can be used to filter everything targeting the
stable branches, both release-XX.YY or staging-XX.YY.
- "4.workflow: package set updates" can be used to filter the
haskell/python package set PRs. Can be combined with "6.topic: haskell"
or "6.topic: python".
- "4.workflow: staging" can be used to filter for all staging-next PRs,
both to master and release branches.

(cherry picked from commit 5ed0205f6a9f6b1939c5b9caf835c368b578fb62)

wolfgangwalther committed:
workflows/labeler: fix repo owner condition

Introduced in #402332, but broken on all ends:
- pull_request needs to be event.pull_request
- pull_request.head is an object, not a string
- github.ref_protected is about the target branch, because this runs as
a pull_request_target event

Thus, we need to list the branches manually.

(cherry picked from commit e67623596d0f31f912652729993ec20f8bd91059)

wolfgangwalther committed:
workflows/backport: only trigger on backport labels

The backport job was previously triggered on any label activity, i.e.
when the backport label was already set and a new label added, the
backport would have been triggered again.

That's because github.event_name is actually "pull_request_target" in
this case, not "closed" or "labeled" (the event's types). Thus, this
part of the condition was always true.

This also means that the second part, the startsWith, was never
evaluated. It had its arguments flipped and would have always been
false.

This was introduced in #126825, but has never really worked as intended.

Resolves #199556

(cherry picked from commit b62d9a22fbab05be9b6fdb473d89fc2fd9fe4c1d)

wolfgangwalther committed:
workflows/backport: add "has: port to stable" label on success

This allows filtering for PRs with a backport label, but without the
"has: port to stable" label to find those which need to be manually
acted on.

Resolves #325359

(cherry picked from commit fa154d184f6aa13691ac3c9de7452c3be60844d3)

Mic92 commented on .github/workflows/labels.yml:
```diff
@@ -20,6 +20,23 @@ jobs:
     if: "github.repository_owner == 'NixOS' && !contains(github.event.pull_request.title, '[skip treewide]')"
     steps:
       - uses: actions/labeler@8558fd74291d67161a8a78ce36a881fa63b766a9 # v5.0.0
+        if: "!(github.pull_request.head.repo == 'NixOS' && github.ref_protected)"
```
Do we also need https://github.com/NixOS/nixpkgs/pull/404152/files ?

wolfgangwalther committed:
workflows/backport: fix conditional

(cherry picked from commit b6375b21c0996de13051110c80ff72ec1669261c)

wolfgangwalther committed:
workflows/backport: avoid broken korthout/backport-action output

Somehow, the was_successful output didn't work correctly. It's hard to
imagine that the created_pull_numbers output fails the same way,
because... when the backport action fails there **are no pull request
numbers**.

(cherry picked from commit 2480e356dc9309a48389aacb80f4bdd8fd2a0897)

wolfgangwalther committed:
workflows/labeler: fix double quotes

Because GitHub Actions doesn't like double quotes.

(cherry picked from commit 6cb93f221b91b747962cae9d1de26d99f4e1003c)

wolfgangwalther committed:
workflows/labels: fix wrong yaml file

(cherry picked from commit de670fa46645033491f265417047ecdeac648bab)
